src/eradiate_plugins/bsdfs/rtls.cpp

```diff
@@ -0,0 +1,322 @@
+
+        std::ostringstream oss;
+        oss << "Geometrical setup" << std::endl
+            << "  phi_i: " << dr::asin(sin_phi_i) << std::endl
+            << "    sin(phi_i): " << sin_phi_i << std::endl
+            << "    cos(phi_i): " << cos_phi_i << std::endl
+            << "  phi_o: " << dr::asin(sin_phi_o) << std::endl
+            << "    sin(phi_o): " << sin_phi_o << std::endl
+            << "    cos(phi_o): " << cos_phi_o << std::endl
+            << "  d_phi: " << dr::asin(sin_d_phi) << std::endl
+            << "    sin(d_phi): " << sin_d_phi << std::endl
+            << "    cos(d_phi): " << cos_d_phi << std::endl;
+        Log(Trace, oss.str().c_str());
+
+        const Float cos_psi =
+            cos_theta_i * cos_theta_o + sin_theta_i * sin_theta_o * cos_d_phi;
+
+        const Float sin_psi = dr::sqrt(1 - dr::sqr(cos_psi));
+        const Float psi     = dr::acos(cos_psi);
+
+        const UnpolarizedSpectrum K_iso = eval_K_iso();
+        const UnpolarizedSpectrum K_vol =
+            eval_K_vol(cos_theta_i, cos_theta_o, cos_psi, sin_psi, psi);
+
+        UnpolarizedSpectrum K_geo;
+        if (unlikely(m_r - m_b > dr::Epsilon<ScalarFloat>)) {
+            Log(Debug, "Using different b and r values forcing extra angles "
+                       "calculations");
+
+            const Float tan_theta_i_prim = m_b / m_r * tan_theta_i;
+            const Float tan_theta_o_prim = m_b / m_r * tan_theta_o;
+            const Float theta_i_prim     = dr::atan(tan_theta_i_prim);
+            const Float theta_o_prim     = dr::atan(tan_theta_o_prim);
+            const Float cos_theta_i_prim = dr::cos(theta_i_prim);
+            const Float cos_theta_o_prim = dr::cos(theta_o_prim);
+            const Float sin_theta_i_prim = dr::sin(theta_i_prim);
+            const Float sin_theta_o_prim = dr::sin(theta_o_prim);
+            const Float cos_psi_prim =
+                cos_theta_i_prim * cos_theta_o_prim +
+                sin_theta_i_prim * sin_theta_o_prim * cos_d_phi;
+
+            K_geo = eval_K_geo(cos_theta_i_prim, cos_theta_o_prim,
+                               tan_theta_i_prim, tan_theta_o_prim, cos_d_phi,
+                               sin_d_phi, cos_psi_prim);
+        } else {
+            Log(Trace, "Using similar b and r values, skipping extra angles "
+                       "calculations");
+            K_geo = eval_K_geo(cos_theta_i, cos_theta_o, tan_theta_i,
+                               tan_theta_o, cos_d_phi, sin_d_phi, cos_psi);
+        }
+
+        oss = std::ostringstream();
+        oss << "Results" << std::endl
+            << "  K_iso: " << K_iso << std::endl
+            << "  K_vol:" << K_vol << std::endl
+            << "  K_geo: " << K_geo << std::endl;
+        Log(Trace, oss.str().c_str());
+
+        const UnpolarizedSpectrum value =
+            (f_iso * K_iso + f_vol * K_vol + f_geo * K_geo);
+
+        return depolarizer<Spectrum>(dr::select(value > 0.0, value, dr::NaN<ScalarFloat>));
```

Comment:
Documenting why you return NaN would be appropriate, this will invalidate the sample and might have harmful consequences on the whole radiance sample.